### PR TITLE
feat: filter out nodes in different cluster session

### PR DESCRIPTION
**Pull request by gclaramunt**
Filter out nodes in a different session than seed hosts. Requires https://github.com/Constellation-Labs/tessellation/pull/976

### CODE REVIEW

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 1]
```diff
@@ -68,6 +69,7 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
                   )
                   .getOrElse(IO.pure(currentStatus))
             )
+            .flatMap (selectSameSessionNodes(hosts))
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: formatting

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 2]
```diff
@@ -173,8 +175,9 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
     SortedSet[Addr]() ++ clusterInfo.toNel.foldLeft(List.empty[Addr])(
       (acc, bcc) =>
         bcc match {
-          case (_, Some(hosts: List[Info])) => acc ++ hosts.map(n => Addr(n.ip, n.publicPort)).filterNot(clusterInfo(_).isDefined)
-          case _                            => acc
+          case (_, Some(hosts: List[Info])) =>
+            acc ++ hosts.map(n => Addr(n.ip, n.publicPort)).filterNot(clusterInfo(_).isDefined)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: I don't see any change here 🤔

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 3]
```diff
@@ -231,3 +234,23 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
       )
   }
 }
+
+object Manager {
+
+  val logger: Logger[IO] = Slf4jLogger.getLogger[IO]
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: get logger for class

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 4]
```diff
@@ -231,3 +234,23 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
...
+    val (sameSessionNodes, otherSessionNodes) = currentHosts.toSortedMap.partition {
+      case (_, v) =>
+        knownSession.flatMap(session => v.map(_.forall(_.clusterSession == session))).getOrElse(true)
+    }
+
+    Applicative[IO].whenA(otherSessionNodes.nonEmpty) {
+      logger.warn(
+        s"Evicted nodes not in cluster session ${knownSession.getOrElse("unknown")}: ${otherSessionNodes.keySet} "
+      )
+    } *> IO.pure(NonEmptyMap.fromMapUnsafe(sameSessionNodes))
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: please use flatMap over product

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: Changed to flatMap alias (>>)

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 3 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I've actually, got rid of the logger here, and split the code into two functions, leaving the logging in the main calling method

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 5]
```diff
@@ -68,6 +70,15 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
...
                   .getOrElse(IO.pure(currentStatus))
             )
+            .flatMap { currentNodes =>
+              val session = findSession(hosts)
+              val (sameSessionNodes, otherSessionNodes) = selectSameSessionNodes(session, currentNodes)
+              Applicative[IO].whenA(otherSessionNodes.nonEmpty) {
+                logger.warn(
+                  s"Evicted nodes not in cluster session ${session.getOrElse("unknown")}: ${otherSessionNodes.keySet} "
+                )
+              } >> IO.pure(NonEmptyMap.fromMapUnsafe(sameSessionNodes))
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: hmm, is it possible that `sameSessionNodes` would be an empty one? if so, it's indeed unsafe to call `fromMapUnsafe`

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 6]
```diff
@@ -231,3 +242,18 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
       )
   }
 }
+
+object Manager {
+
+  def selectSameSessionNodes(knownSession: Option[Long], currentHosts: NonEmptyMap[Addr, Option[List[Info]]]): (SortedMap[Addr, Option[List[Info]]], SortedMap[Addr, Option[List[Info]]])  = {
+    val (sameSessionNodes, otherSessionNodes) = currentHosts.toSortedMap.partition {
+      case (_, v) =>
+        knownSession.flatMap(session => v.map(_.forall(_.clusterSession == session))).getOrElse(true)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: please use cats eqv to compare sessions

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 7]
```diff
@@ -231,3 +242,18 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
       )
   }
 }
+
+object Manager {
+
+  def selectSameSessionNodes(knownSession: Option[Long], currentHosts: NonEmptyMap[Addr, Option[List[Info]]]): (SortedMap[Addr, Option[List[Info]]], SortedMap[Addr, Option[List[Info]]])  = {
+    val (sameSessionNodes, otherSessionNodes) = currentHosts.toSortedMap.partition {
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: this is exactly what you return as far as I see, so you don't need to create an additional `val`, do you? 🤔

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 6 above)
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: why `getOrElse` has default `true`? 🤔

File: src/test/scala/org/constellation/lb/ManagerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import org.http4s.implicits.http4sLiteralsSyntax
+import org.scalatest.{FunSpec, Matchers}
+
+import java.net.InetAddress
+
+class ManagerSpec extends FunSpec with Matchers {
+
+  describe("findSession"){
+    it("find a session from a nodes map") {
+      val nodes =NonEmptyMap.of(
```
Comment (marcinwadon) [inline src/test/scala/org/constellation/lb/ManagerSpec.scala]: format

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 5 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: is not possible, at least it should have the seed host nodes that are the sources of the session used to filter,

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: hm, but it can remove them later on, right? what if non of the peers are in the majority? it may be an empty list then 🤔

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: source nodes are being used for the very first scan. then, they are like every other node. if they would go down, the load balancer may throw an error here 🤔

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 7 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: not really, but I thought it was clearer to document which collection is which, but maybe it makes more sense to move to the method's scala doc.

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 6 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala:251]: true or false, it doesn't matter, it means knownSession is None, which is actually a pathological case (it should never happen) . Since there's no session to discriminate, all values of the partition will either fall on the first collection (if we default to true) or in the second collection (if we default to false)

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 5 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: if the node doesn't respond, is kept in the map with value set to None, so the map still won't be empty.

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 6 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: Actually, it matters, it needs to be true: if all the nodes go down, we won't find any session, but instead of filter out everything, it just keeps the collection unchanged (if we set to false, we'll remove all the addresses and the load balancer won't be able to recover)

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: Hm, I got it. then, maybe it's not worth to have `Option[Long]` but just `Long` if we should always know the session. otherwise, it's a fetch or unexpected error and we should not proceed

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 7 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: ok then :+1: works for me

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 5 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: yes, but at some point we remove nodes that are below the threshold. please check that flow. otherwise I'm fine

edit: I remember from logs from the past that we sometimes removed all the nodes and we ended up with an empty list of nodes

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 6 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: Hmmm. If all nodes go down we probably should go back to source nodes again. Otherwise load balancer may stuck in a fork

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 5 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I don't see any place where keys are removed form the map (also, any removal operation change the type to SortedMap, which would mean an unsafe  conversion back to NeMap)

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 6 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: yeah, that's actually an argument for always consider seed nodes only

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 5 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: What happens when the node is below threshold? Don't we remove it? Or we just keep it but does not use it for proxy?

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: That's taken care downstream (but it doesn't actually evict): 

buildClusterStatus(status).flatMap {
                  case (activeHosts, inactiveHosts) =>
                    val clusterHosts = (activeHosts ++ inactiveHosts) // TODO: Add Eviction of inactive hosts

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 9]
```diff
@@ -68,6 +71,15 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
                   )
                   .getOrElse(IO.pure(currentStatus))
             )
+            .flatMap { currentNodes =>
+              val session = findSession(hosts.)
```
Comment (kpudlik) [inline src/main/scala/org/constellation/lb/Manager.scala]: It won't compile. Have you compiled the project before pushing?

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 10]
```diff
@@ -231,3 +236,39 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
...
+    currentHosts.toSortedMap.partition { case (_, v) => knownSession.flatMap(session => v.map(_.forall(_.clusterSession === session))).getOrElse(true) }
+
+  /**
+   * Finds the first cluster session from a map of address -> nodes
+   *
+   * @param hosts address to nodes map
+   * @return cluster session id found if any
+   */
+  def findSession(hosts: NonEmptyMap[Addr, Option[List[Info]]]): Option[Long] =
+    hosts.find(_.exists(_.nonEmpty)).flatMap(_.flatMap(_.headOption).map(_.clusterSession))
```
Comment (kpudlik) [inline src/main/scala/org/constellation/lb/Manager.scala]: Can be simplified in a single `collectFirst` call. There is no need to use `headOption` since you already filtered the `empty` ones:

```scala
hosts.toSortedMap.collectFirst {  case (_, Some(info :: _) => info.clusterSession }
```

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I like it

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 6 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I've modified the logic, if the session cannot be determined (seed nodes are down) no discovery is performed, i,e, no new nodes are added.

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 11]
```diff
@@ -55,179 +59,235 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
...
-            _ =>
-              BlazeBuilder[IO]
-                .bindHttp(config.settingsPort, config.`if`)
-                .mountService(utilsHealth, "/utils")
-                .mountService(settingsRoutes, "/settings")
-                .serve
-                .compile
-                .drain
+def getClusterInfo(
+                    hosts: NonEmptySet[Addr]
```
Comment (kpudlik) [inline src/main/scala/org/constellation/lb/Manager.scala]: I guess you need to configure the code formatter properly. The project uses https://scalameta.org/scalafmt/

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: dang, intelliJ is not picking it pu :( will fix it

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: ran scalafmt from command line :)

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: @gclaramunt looks like the formatter that you've used made much more changes that ones required for your change :| is it possible that you commit reformatted code (with no more logic changes) as a separate commit, and your changes as a separate commit? right now it's hard to distinguish what is the effect of the formatter and what is really a change needed here.

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I've reverted back the format changes, even if there are some now, they're are related to the logic change (minus removing some unused imports)

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 5 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: as per the discussion, the code now finds and keeps only one session, which is reset on load balancer reset.

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 11 above)
Comment (kpudlik) [reply src/main/scala/org/constellation/lb/Manager.scala]: image

Just make sure that you have it set to Scalafmt instead of IntelliJ @gclaramunt

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 12]
```diff
@@ -38,48 +42,74 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
   private lazy val hostsRef =
     Ref.unsafe[IO, NonEmptyMap[Addr, Option[List[Info]]]](init.map(addr => addr -> None).toNem)
 
+    private lazy val sessionRef =  Ref.unsafe[IO, Option[Long]](None)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: Looks like you have reformatted code that you haven't touched but here for example is formatted wildly (2 spaces after `=`)? If you want to reformat the code using scalafmt or anything, would be great to make it as a totally separate commit. Otherwise, it's hard to figure out what is a real change here

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 13]
```diff
@@ -38,48 +42,74 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
...
+      .flatMap {
+       hosts =>
+         sessionRef.get.flatMap {
+           case oSession@Some(_) => IO.pure((hosts, oSession))
+           case None =>
+             val session = findSession(filterKeys(init.toNes, hosts))
+             sessionRef.set(session).map(_ => (hosts, session))
+         }
+      }
+      z.flatMap { case (hosts, session) =>
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: I think you don't need this `z` variable

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: yeah, that type debugging step.

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 12 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: fixed with scalafmt

File: src/it/scala/constellation/IntegrationSpec.scala [hunk 14]
```diff
@@ -1,99 +1,103 @@
...
-  val clusterInit = fakeSetup.map(i => s"localhost:${i.publicPort}").toList
 
-  val nodes =
-    fakeSetup
-      .map(i => new FakeNode(i.publicPort, i.p2pPort, fakeSetup.toList, Json.obj("node-id" -> i.id.hex.asJson)))
-      .map(_.run)
-      .parSequence
+  def nodesServices(nodesInfo: List[Info]) =
+    nodesInfo
+      .map{n => println(s"Starting for $n with $nodesInfo"); n}
```
Comment (marcinwadon) [inline src/it/scala/constellation/IntegrationSpec.scala]: `println` is a side effect. you can use `Console` from cats and use `flatTap` operator

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 15]
```diff
@@ -38,48 +44,80 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
...
-                      .foldLeft(status)((acc, addr) => acc.add(addr, Option.empty[List[Info]]))
-
-                    hostsRef.set(clusterHosts).flatTap(_ => updateLbSetup(activeHosts))
+            .flatMap { currentStatus =>
+              sessionRef.get
+                .flatMap {
+                  case oSession @ Some(_) => IO.pure(oSession)
+                  case None =>
+                    val session = findSession(filterKeys(init.toNes, currentStatus))
+                    sessionRef.set(session).map(_ => session)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: `.map(_ => a)` => `.as(a)`

File: src/main/scala/org/constellation/lb/Manager.scala [hunk 16]
```diff
@@ -231,3 +256,66 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
       )
   }
 }
+
+object Manager {
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/lb/Manager.scala]: why those fns in object?

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: They don't depend on the object's state so you don't really need  an instance to use them.

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: I see, but they are like internal for the manager (mostly?), so imo they should be private fns inside

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 15 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: Any preference between 
`sessionRef.modify( _ => (session, session))` vs `sessionRef.set(session).as(session)`

neither are particularly satisfying

File: src/it/scala/constellation/IntegrationSpec.scala (hunk 14 above)
Comment (gclaramunt) [reply src/it/scala/constellation/IntegrationSpec.scala]: leftover code, is not needed, I'll remove it.

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 15 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: if we had a generic `def dup[T](t: T): (T, T) = (t, t)` elsewhere, we can do `sessionRef.modify( _ => dup(session))`
Or even better, with `def constTuple[T](x: T)(y: T): (T, T) = (x, x)` we end with  `sessionRef.modify(constTuple(session))`

File: src/main/scala/org/constellation/lb/Manager.scala (hunk 16 above)
Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: filterKeys is generic enough to even be in a separate lib, the others can be private, but then, unit test them would be almost impossible.  I like to separate the pure functions from the stateful object in a separate module, but happy to move to private fn if is our standard.  
Another option is to make all the functions totally generic in `K` and `T` in `NonEmptyMap[K, Option[NonEmptyList[T]]]` and move to a totally different module.

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: if you need internal functions to make tests, it's considered "test smells" meaning most likely you're doing something wrong.

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I don't need the functions to make the tests,  is the other way around, I want to test those functions. Essentially the question is "do we want finer grained unit tests?"

Comment (marcinwadon) [reply src/main/scala/org/constellation/lb/Manager.scala]: if so, they should be kinda external, not connected with the Manager itself

Comment (gclaramunt) [reply src/main/scala/org/constellation/lb/Manager.scala]: I've ended up generalizing the functions and moving them to another module

**Review by marcinwadon — APPROVED** [review]
